### ISSUE TITLE
Multiple Partial Overlap Bug

### ISSUE THREAD

**Issue by Vatvo** [Logicalshift/flo_curves#30]
Hello Flo_Curves Devs,

I found a bug with the marking of interior/exterior edges when subtracting complex shapes in a specific way in both v0.7.3 and v0.8.0. 

It begins with a base shape created by merging four SimpleBezierPaths into a GraphPath:
<img width="1101" alt="Screenshot 2024-06-26 at 1 24 22 PM" src="https://github.com/Logicalshift/flo_curves/assets/34244664/8f69c0af-564d-4a8a-8cbf-be2e8732b204">

And a cutout GraphPath like this:
<img width="739" alt="Screenshot 2024-06-26 at 1 28 40 PM" src="https://github.com/Logicalshift/flo_curves/assets/34244664/5f76e716-03ba-4373-a5c5-b61b77bcb5a1">

The cutout path is then subtracted from the base shape using GraphPath::collide() and then GraphPath::set_exterior_by_subtracting(). This produces the incorrect path seen here:
<img width="1101" alt="Noncomposite_Incorrect" src="https://github.com/Logicalshift/flo_curves/assets/34244664/4080cf81-871d-4d73-9340-428d09cba432">

Here is a recreation of what it should look like if it were correct:
<img width="1101" alt="Composite_Correct" src="https://github.com/Logicalshift/flo_curves/assets/34244664/720b6fe5-1d0d-4b7c-b7b4-fe7f3c4d3f20">

The result of this error (when healed using the heal_exterior_gaps() function) looks like this:
![Thin_Lock_Broken](https://github.com/Logicalshift/flo_curves/assets/34244664/480a9f56-63d0-470d-9626-fabf3934a6f7)

When it should look like this:
![Thin_Lock_Fixed](https://github.com/Logicalshift/flo_curves/assets/34244664/8a212df7-5316-4617-a937-6841a7214344)

The two black and white images were made using the original icons with which this bug was discovered, but I have written a test using the more minimal reproduction seen in the debug images.

The inner loop seen in the top right of the debug image seems to be important. When it is removed, the bug does not occur. 
<img width="1101" alt="pathloop" src="https://github.com/Logicalshift/flo_curves/assets/34244664/f87314db-4a2c-4b1a-8bb6-de1220e33c24">

Also, if the top of the cutout shape, which intersects that loop, is moved higher (to around y=100) the bug does not occur.
<img width="1101" alt="path100" src="https://github.com/Logicalshift/flo_curves/assets/34244664/cd4cccbb-2f57-4da2-8df3-60101772ee7c">

When that point is at its normal position of y=160 it breaks.
<img width="1101" alt="path160" src="https://github.com/Logicalshift/flo_curves/assets/34244664/645630c5-8074-4d71-a3d9-9045d996b56e">

I believe the bug is somewhere in the set_exterior_by_subtracting() function, or the function it calls, though I was unable to locate exactly where.

The subtract_multiple_partial_overlap() test I have proposed in pull request #29 creates these shapes, subtracts them, locates the edges of the inner arch, and checks their interior/exterior statuses. If they are correctly labelled as exterior, it passes. I believe that if this test passes, the bug would be fixed.

Also, I’m not sure if this is relevant, but maybe it will be useful information somehow. The edge indicated in the following image is actually two lines that partially overlap:
<img width="1101" alt="twolines02" src="https://github.com/Logicalshift/flo_curves/assets/34244664/e6b39779-ac6d-4cb3-88ba-e414561d17ad">

I would think that Flo_Curves should see the overlap and split the edge up appropriately.
<img width="1101" alt="splitlines02" src="https://github.com/Logicalshift/flo_curves/assets/34244664/ec383bc7-2ddd-4883-9f50-280daa80c4c3">

I could be wrong about what this is supposed to look like, but maybe this information will be useful.

**Comment by Logicalshift** [comment]
Hi, thanks for the comprehensive report, I was easily able to reproduce the bug. I've got some first impressions for now, I should be able to try to figure out how to fix this this weekend.

I believe you've also identified the cause here: any overlapping lines should have caused a collision so there shouldn't be any of differing lengths. That will always cause incorrect behaviour, but precisely what happens depends on the ray that intersects the lines, hence why altering the shape seems to change the outcome. The longer line is inside the shape at the top, but outside at the bottom, so there are two possible results.

(Partially overlapping curves are the worst for figuring out collisions: apparently even when they are also lines)

There are some other possibilities: rays that run nearly exactly along a linear edge can cause glitches too, and rays that cross two overlapping edges have previously caused problems (v0.8 rewrote the algorithm for that last one so I'm hoping that's actually fully fixed now), and there's always the unexpected wildcard bug in some other part of the code.

I'll figure this out for sure this weekend, but this analysis has been really useful in narrowing down what could be going wrong.

**Comment by Logicalshift** [comment]
I believe I've fixed this as of d425e83f91503f9af6f363058dab3dd835fca432 which is on the v0.8 branch.

There were a couple of problems: the main one was to do with overlapping curves. If you had a situation like this, where two overlapping lines are moving in opposite directions and the end points only :

```
   *------>
       <-------*
```

flo_curves would not recognise them as overlapping. This was the main reason the intersection wasn't being generated.

Then there was a filter being applied that was supposed to stop the last point of one path edge from intersecting with the first point of the next that *also* removed the intersection. That only makes sense when computing self-intersections so I've removed it.

Finally fixing the overlapping bug introduced a new bug where very small subdivided curves would be seen as overlapping, which would generate extra collisions. I've fixed that with a check that discards overlaps that are very close to the start and end of the curve (the points should snap together later on anyway)

**Comment by Vatvo** [comment]
Looks good. Thanks for the quick response. If it's not too much trouble, could the fix be added to the 0.7 branch too?

**Comment by Logicalshift** [comment]
Yep, no problem. I've copied the fix over into the 0.7 branch and released it with v0.7.3
